### PR TITLE
[BEAM-10669] Add support for Dataflow Templates

### PR DESCRIPTION
**Pull request by kkucharc**
…ved most of functionalities to expand methods

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
Build python source distribution and wheels

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/crosslanguage/SnowflakeReadRegistrar.java [hunk 1]
```diff
@@ -56,11 +54,9 @@ public ReadBuilder() {}
 
     @Override
     public PTransform<PBegin, PCollection<byte[]>> buildExternal(ReadConfiguration c) {
-      SnowflakeCredentials credentials = SnowflakeCredentialsFactory.of(c);
-
       SerializableFunction<Void, DataSource> dataSourceSerializableFunction =
           SnowflakeIO.DataSourceProviderFromDataSourceConfiguration.of(
-              SnowflakeIO.DataSourceConfiguration.create(credentials)
```
Comment (kkucharc) [inline sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/crosslanguage/SnowflakeReadRegistrar.java:63]: @purbanow Do you know how to test it? Maybe I should add credentials in the other way?

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java [hunk 2]
```diff
@@ -457,22 +474,22 @@ private String makeTmpDirName() {
...
+      private final ValueProvider<String> database;
+      private final ValueProvider<String> schema;
+      private final ValueProvider<String> table;
+      private final ValueProvider<String> storageIntegrationName;
+      private final ValueProvider<String> stagingBucketDir;
       private final String tmpDirName;
-      private final String storageIntegrationName;
-      private final String stagingBucketDir;
       private final SnowflakeService snowflakeService;
       private final String quotationMark;
```
Comment (purbanow) [inline sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java]: Can we pass `quotationMark` as option ? Maybe is worth to make this var supported at runtime ? We can make Issue for this if currently there is no time for this kind of improvements.

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/crosslanguage/SnowflakeReadRegistrar.java (hunk 1 above)
Comment (purbanow) [reply sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/crosslanguage/SnowflakeReadRegistrar.java:63]: i think in the past we had tests for cross-language using java. I think @piotr-szuberski wrote those tests, but we decided to drop them.

File: sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/TestSnowflakePipelineOptions.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+  void setPortNumber(String portNumber);
+
+  @Description("Login timeout. Optional.")
+  @Default.String("")
+  String getLoginTimeout();
+
+  void setLoginTimeout(String loginTimeout);
+
+  @Description("Temporary GCS bucket name.")
+  String getStagingBucketName();
```
Comment (purbanow) [inline sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/TestSnowflakePipelineOptions.java:136]: Maybe we can add @Validation.Required ?

File: sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/TestSnowflakePipelineOptions.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+  void setStagingBucketName(String stagingBucketName);
+
+  @Description("Storage integration name")
+  String getStorageIntegrationName();
```
Comment (purbanow) [inline sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/TestSnowflakePipelineOptions.java:142]: Same here with @Validation.Required ?

File: sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/TestSnowflakePipelineOptions.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+  @Description("Warehouse to use. Optional.")
+  @Default.String("")
+  String getWarehouse();
+
+  void setWarehouse(String warehouse);
+
+  @Description("Database name to connect to. Optional.")
+  @Default.String("")
+  String getDatabase();
```
Comment (purbanow) [inline sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/TestSnowflakePipelineOptions.java:88]: Maybe we can add @Validation.Required?

**Review by purbanow — APPROVED** [review]

**Comment by purbanow** [comment]
LGTM with small comments :)

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java (hunk 2 above)
Comment (kkucharc) [reply sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java]: Sure

Comment (kkucharc) [reply sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java]: And done

File: sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/TestSnowflakePipelineOptions.java (hunk 3 above)
Comment (kkucharc) [reply sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/TestSnowflakePipelineOptions.java:136]: Sure, thanks! Added

File: sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/TestSnowflakePipelineOptions.java (hunk 4 above)
Comment (kkucharc) [reply sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/TestSnowflakePipelineOptions.java:142]: Sure, thanks! Added

File: sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/TestSnowflakePipelineOptions.java (hunk 5 above)
Comment (kkucharc) [reply sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/TestSnowflakePipelineOptions.java:88]: Sure, thanks! Added

**Pull request closed without merging**